testsuite/tests/basic/division_by_constant.ml

```diff
@@ -0,0 +1,366 @@
+(* TEST
+   { native; }
+*)
+open Nativeint
+
+let test_cases =
+  let int_size = Nativeint.size in
+  let rng = Random.State.make [| int_size |] in
+  (* sparse test cases, concentrated around 0, the limits, and powers of 2 *)
+  List.init (int_size - 1) (fun size ->
+    let bit = (shift_left one) size in
+    let rand () =
+      Random.State.nativeint_in_range rng
+        ~min:(shift_right_logical bit 1)
+        ~max:(pred bit)
+    in
+    [ bit;
+      rand ();
+      lognot (rand ());
+      (sub max_int) (rand ());
+      lognot ((sub max_int) (rand ())) ])
+  |> List.concat
+  |> List.sort_uniq compare
+
+let[@inline never] check ~dividend ~divisor ~quotient ~remainder =
+  assert (divisor <> 0n);
+  assert (quotient = div dividend divisor);
+  assert (remainder = rem dividend divisor);
+  assert (dividend = add (mul quotient divisor) remainder);
+  assert
+    (if divisor = min_int
+     then remainder <> min_int
+     else abs remainder < abs divisor);
+  let[@inline] sign x = if x < 0n then -1 else if x = 0n then 0 else 1 in
+  assert (remainder = 0n || sign remainder = sign dividend)
+
+let[@inline always] [@specialise always] check ~divisor =
+  ListLabels.iter test_cases ~f:(fun dividend ->
+    (Sys.opaque_identity check)
+      ~dividend
+      ~divisor
+      ~quotient:(div dividend divisor)
+      ~remainder:(rem dividend divisor)
+  )
+
+(* manual tests *)
+let () = check ~divisor:3n
+let () = check ~divisor:(-3n)
+let () = check ~divisor:65537n
+let () = check ~divisor:(-65537n)
+
+(* checks generated by test_cases. These get inlined so the division by a constant is
+   replaced *)
```

Comment:
this comment is quite confusing:
- you could specify that it was generated by a manual run of the `test_cases` function to have a hard-coded list of semi-random numbers (whereas all divisors are tested on the list generated by `test_cases`)
- the comment about `check` being inlined also applies to the "manual" tests above, so stating it here is confusing. You should instead add a comment to this effect on the definition of the `check` function explaining how the `[@inline]` annotations are used to ensure that the divisor is a compile-time known constant but not the dividend. That would also help specify exactly what is being tested by this file.